pkg/ccl/backupccl/restore_old_versions_test.go

```diff
@@ -730,3 +751,48 @@ func TestRestoreWithDroppedSchemaCorruption(t *testing.T) {
 	}
 	require.Falsef(t, hasSameNameSchema(dbName), "corrupted descriptor exists")
 }
+
+// restorePublicSchemaRemap tests that if we're restoring a database from
+// an older version where the database has a synthetic public schema, a real
+// descriptor backed public schema is created and the tables in the schema
+// are correctly mapped to the new public schema.
+func restorePublicSchemaRemap(exportDir string) func(t *testing.T) {
+	return func(t *testing.T) {
+		const numAccounts = 1000
+		_, _, _, tmpDir, cleanupFn := BackupRestoreTestSetup(t, MultiNode, numAccounts, InitManualReplication)
+		defer cleanupFn()
+
+		_, _, sqlDB, cleanup := backupRestoreTestSetupEmpty(t, singleNode, tmpDir,
+			InitManualReplication, base.TestClusterArgs{})
+		defer cleanup()
+		err := os.Symlink(exportDir, filepath.Join(tmpDir, "foo"))
+		require.NoError(t, err)
+
+		sqlDB.Exec(t, fmt.Sprintf("RESTORE DATABASE d FROM '%s'", LocalFoo))
+
+		var restoredDBID, publicSchemaID int
+		row := sqlDB.QueryRow(t, `SELECT id FROM system.namespace WHERE name='d' AND "parentID"=0`)
+		row.Scan(&restoredDBID)
+		row = sqlDB.QueryRow(t, fmt.Sprintf(`SELECT id FROM system.namespace WHERE name='public' AND "parentID"=%d`, restoredDBID))
+		row.Scan(&publicSchemaID)
+
+		if publicSchemaID == keys.PublicSchemaID {
+			t.Fatalf("expected public schema id to not be %d", keys.PublicSchemaID)
+		}
+
+		row = sqlDB.QueryRow(t,
+			fmt.Sprintf(`SELECT count(1) FROM system.namespace WHERE name='t' AND "parentID"=%d AND "parentSchemaID"=%d`, restoredDBID, publicSchemaID))
+		require.NotNil(t, row)
+
+		sqlDB.CheckQueryResults(t, `SELECT x FROM d.s.t`, [][]string{{"1"}, {"2"}})
+		sqlDB.CheckQueryResults(t, `SELECT x FROM d.public.t`, [][]string{{"3"}, {"4"}})
+
+		// Test restoring a single table and ensuring that d.public.t which
+		// previously had a synthetic public schema gets correctly restored into the
+		// descriptor backed public schema of database test.
+		sqlDB.Exec(t, `CREATE DATABASE test`)
+		sqlDB.Exec(t, `RESTORE d.public.t FROM $1 WITH into_db = 'test'`, LocalFoo)
+
```

Comment:
Maybe I'm being pedantic but can we check the public desc ID before restore for the database test, and then check that the parentSchemaID of the table is that. I would assume the SELECT wouldnt work if this was busted but 🤷 if you don't mind.

Reply:
Added

```
row = sqlDB.QueryRow(t, `SELECT id FROM system.namespace WHERE name='test' AND "parentID"=0`)
var parentDBID int
row.Scan(&parentDBID)

row = sqlDB.QueryRow(t, fmt.Sprintf(`SELECT id FROM system.namespace WHERE name='public' AND "parentID"=%d`, parentDBID))
row.Scan(&publicSchemaID)

if publicSchemaID == keys.PublicSchemaID || publicSchemaID == int(descpb.InvalidID) {
	t.Errorf(fmt.Sprintf("expected public schema id to not be %d or %d, found %d", keys.PublicSchemaID, descpb.InvalidID, publicSchemaID))
}
```